src/main/cpp/Elevator/FeedforwardPID.cpp

```diff
@@ -0,0 +1,272 @@
+    frc::SmartDashboard::PutNumber("expected ev position", expected_pose.position);
+    frc::SmartDashboard::PutNumber("expected acceleration", expected_pose.acceleration);
+
+    frc::SmartDashboard::PutNumber("position error", expected_pose.position - current_values.position);
+    frc::SmartDashboard::PutNumber("velocity error", expected_pose.velocity - current_values.velocity);
+    frc::SmartDashboard::PutNumber("acceleration error", expected_pose.acceleration - current_values.acceleration);
+
+    frc::SmartDashboard::PutNumber("ff voltage", feedforward_voltage);
+    frc::SmartDashboard::PutNumber("pid voltage", pid_voltage);
+
+    return feedforward_voltage + pid_voltage;
+}
+
+/**
+ * @brief PID calculations to make feedforward loop more accurate
+ *
+ * @param expected Pose containing information about where system should be
+ * @param current Pose containing information about where system actually is
+ * @return double voltage to add to feedforward loop to compensate for inaccuracies in velocity/position.
+ */
+double FeedforwardPID::calculatePIDVoltage(Poses::Pose1D expected, Poses::Pose1D current)
+{
+    return kp * (expected.velocity - current.velocity) + kd * (expected.position - current.position);
+}
+
+/**
+ * @brief Returns the voltage based on the feedforward formula
+ *
+ * @param velocity - expected velocity, based on motion profile
+ * @param acceleration - expected acceleration, based on motion profile
+ *
+ * @return the voltage to move
+ */
+double FeedforwardPID::calculateFeedforwardVoltage(double velocity, double acceleration)
+{
+    return ks * sign(velocity) + kg + kv * velocity + ka * acceleration;
+}
+
+/**
+ * @brief The sign of a double
+ *
+ * @param value the value to get the sign of
+ * @return 1.0 if the value is greater than 0, or -1.0 otherwise
+ */
+double FeedforwardPID::sign(double value)
+{
+    return value > 0.0 ? 1.0 : -1.0;
+}
+
+/**
+ * @brief Starts the timer
+ *
+ */
+void FeedforwardPID::start()
+{
+    timer.Start();
+    isRunning = true;
+}
+
+/**
+ * @brief Stops the timer
+ *  
+ */
+void FeedforwardPID::stop() {
+    timer.Stop();
+}
+
+/**
+ * Resets and starts the timer 
+ * 
+ */
+void FeedforwardPID::reset() {
+    timer.Reset();
+    start();
+}
+
+/**
+ * @brief A feedforward function that gets the elevator pose based on current time
+ *
+ * @param time current system time
+ * @return pose a pose containing the distance
+ */
+Poses::Pose1D FeedforwardPID::getExpectedPose(double time)
+{
+    Poses::Pose1D pose;
+
+    // the time spent accelerating (or decelerating)
+    double acceleration_time = max_velocity / max_acceleration;
+    frc::SmartDashboard::PutNumber("acc time: ", acceleration_time);
+
+    // the time spent maintaining a constant velocity
+    double velocity_time = (max_distance_ - max_velocity * acceleration_time) / max_velocity;
+
+    // velocity should never be negative
+    velocity_time = velocity_time < 0 ? 0.0 : velocity_time;
```

Comment:
You should probably deal with the triangle vs trapezoid profile here with an if statement instead of a ternary expression